File: src/models/palace.rs
use serde::{Deserialize, Serialize};

use crate::data::heavenly_stems::get_heavenly_stem_info;
use crate::data::stars::StarKey;
use crate::data::types::*;
use crate::models::star::Star;

/// 宫位上的大限信息。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Decadal {
    /// 大限起止虚岁（含两端）
    pub range: (u32, u32),
    /// 大限天干
    pub heavenly_stem: HeavenlyStem,
    /// 大限地支
    pub earthly_branch: EarthlyBranch,
}

/// 单个宫位：星耀、干支、四组十二神与大限小限信息。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PalaceData {
    /// 宫位索引（0-11，寅宫为 0）
    pub index: usize,
    /// 宫位名称
    pub name: Palace,
    /// 是否身宫
    pub is_body_palace: bool,
    /// 是否来因宫（宫干与年干相同且不在子丑二宫）
    pub is_original_palace: bool,
    /// 宫位天干
    pub heavenly_stem: HeavenlyStem,
    /// 宫位地支
    pub earthly_branch: EarthlyBranch,
    /// 主星列表（按安放顺序）
    pub major_stars: Vec<Star>,
    /// 辅星列表（按安放顺序）
    pub minor_stars: Vec<Star>,
    /// 杂耀列表（按安放顺序）
    pub adjective_stars: Vec<Star>,
    /// 长生十二神
    pub changsheng12: StarKey,
    /// 博士十二神
    pub boshi12: StarKey,
    /// 将前十二神
    pub jiangqian12: StarKey,
    /// 岁前十二神
    pub suiqian12: StarKey,
    /// 大限信息
    pub decadal: Decadal,
    /// 小限经过的虚岁列表
    pub ages: Vec<u32>,
}

impl PalaceData {
    /// 收集宫位内所有星耀的 StarKey
    fn all_star_keys(&self) -> Vec<StarKey> {
        let mut keys = Vec::new();
        for s in &self.major_stars {
            keys.push(s.key);
        }
        for s in &self.minor_stars {
            keys.push(s.key);
        }
        for s in &self.adjective_stars {
            keys.push(s.key);
        }
        keys
    }

    /// 判断宫位是否包含所有指定星耀
    pub fn has(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().all(|s| all_keys.contains(s))
    }

    /// 判断宫位是否不包含所有指定星耀（全不在才返回 true）
    pub fn not_have(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().all(|s| !all_keys.contains(s))
    }

    /// 判断宫位是否包含任一指定星耀
    pub fn has_one_of(&self, stars: &[StarKey]) -> bool {
        let all_keys = self.all_star_keys();
        stars.iter().any(|s| all_keys.contains(s))
    }

    /// 判断是否有指定四化（检查主星和辅星的 mutagen 字段）
    /// TS: only checks majorStars and minorStars (NOT adjectiveStars)
    pub fn has_mutagen(&self, mutagen: Mutagen) -> bool {
        self.major_stars
            .iter()
            .chain(self.minor_stars.iter())
            .any(|s| s.mutagen == Some(mutagen))
    }

    /// 判断是否没有指定四化
    pub fn not_have_mutagen(&self, mutagen: Mutagen) -> bool {
        !self.has_mutagen(mutagen)
    }

    /// 判断是否空宫（没有 star_type == Major 的主星）
    pub fn is_empty(&self) -> bool {
        !self
            .major_stars
            .iter()
            .any(|s| s.star_type == StarType::Major)
    }

    /// 飞化到目标宫位
    /// 根据本宫天干的四化，判断对应四化星是否在目标宫位中
    pub fn flies_to(&self, target_palace: &PalaceData, mutagen: Mutagen) -> bool {
        let info = get_heavenly_stem_info(self.heavenly_stem);
        let mutagen_index = match mutagen {
            Mutagen::Lu => 0,
            Mutagen::Quan => 1,
            Mutagen::Ke => 2,
            Mutagen::Ji => 3,
        };
        let star_key = info.mutagen[mutagen_index];
        target_palace.has(&[star_key])
    }

    /// 自化判断：本宫天干四化星是否在本宫内
    pub fn self_mutaged(&self, mutagen: Mutagen) -> bool {
        self.flies_to(self, mutagen)
    }

    /// 是否有任一自化
    pub fn self_mutaged_one_of(&self) -> bool {
        [Mutagen::Lu, Mutagen::Quan, Mutagen::Ke, Mutagen::Ji]
            .iter()
            .any(|m| self.self_mutaged(*m))
    }

    /// 是否没有自化（检查所有四化，全部没有才返回 true）
    pub fn not_self_mutaged(&self) -> bool {
        !self.self_mutaged_one_of()
    }

    /// 获取四化飞入的宫位索引：依次为禄、权、科、忌对应的星所在宫位，
    /// 未找到的项为 None
    pub fn mutaged_places(&self, all_palaces: &[PalaceData]) -> Vec<Option<usize>> {
        let info = get_heavenly_stem_info(self.heavenly_stem);
        info.mutagen
            .iter()
            .map(|star_key| {
                all_palaces.iter().find_map(|p| {
                    if p.has(&[*star_key]) {
                        Some(p.index)
                    } else {
                        None
                    }
                })
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use crate::astro::builder::by_solar;
    use crate::data::stars::StarKey;
    use crate::data::types::*;

    fn make_astrolabe() -> crate::models::astrolabe::Astrolabe {
        by_solar(
            "2000-8-16",
            2,
            Gender::Female,
            true,
            Language::ZhCN,
            Config::default(),
        )
        .unwrap()
    }

    #[test]
    fn test_palace_has() {
        let a = make_astrolabe();
        // Find a palace with a known major star and test has()
        let found = a
            .palaces
            .iter()
            .find(|p| p.major_stars.iter().any(|s| s.key == StarKey::ZiweiMaj));
        assert!(found.is_some());
        let p = found.unwrap();
        assert!(p.has(&[StarKey::ZiweiMaj]));
    }

    #[test]
    fn test_palace_not_have() {
        let a = make_astrolabe();
        // The palace with Ziwei should not have all 14 major stars
        let p = a
            .palaces
            .iter()
            .find(|p| p.has(&[StarKey::ZiweiMaj]))
            .unwrap();
        // Ziwei and Pojun can't be in the same palace
        assert!(p.not_have(&[StarKey::PojunMaj]) || p.has(&[StarKey::PojunMaj]));
        // A star not present should return true for not_have
        let not_here: Vec<StarKey> = vec![
            StarKey::ZiweiMaj,
            StarKey::TianjiMaj,
            StarKey::TaiyangMaj,
            StarKey::WuquMaj,
            StarKey::TiantongMaj,
            StarKey::LianzhenMaj,
            StarKey::TianfuMaj,
            StarKey::TaiyinMaj,
            StarKey::TanlangMaj,
            StarKey::JumenMaj,
            StarKey::TianxiangMaj,
            StarKey::TianliangMaj,
            StarKey::QishaMaj,
            StarKey::PojunMaj,
        ]
        .into_iter()
        .filter(|k| !p.has(&[*k]))
        .collect();
        if !not_here.is_empty() {
            assert!(p.not_have(&not_here));
        }
    }

    #[test]
    fn test_palace_has_one_of() {
        let a = make_astrolabe();
        let _p = &a.palaces[0];
        // Should have at least one of all 14 major stars (some palace has at least one)
        let all_majors = [
            StarKey::ZiweiMaj,
            StarKey::TianjiMaj,
            StarKey::TaiyangMaj,
            StarKey::WuquMaj,
            StarKey::TiantongMaj,
            StarKey::LianzhenMaj,
            StarKey::TianfuMaj,
            StarKey::TaiyinMaj,
            StarKey::TanlangMaj,
            StarKey::JumenMaj,
            StarKey::TianxiangMaj,
            StarKey::TianliangMaj,
            StarKey::QishaMaj,
            StarKey::PojunMaj,
        ];
        // At least some palace has one of these
        let any_has = a.palaces.iter().any(|p| p.has_one_of(&all_majors));
        assert!(any_has);
    }

    #[test]
    fn test_palace_has_mutagen() {
        let a = make_astrolabe();
        // At least one palace should have 禄
        let has_lu = a.palaces.iter().any(|p| p.has_mutagen(Mutagen::Lu));
        assert!(has_lu);
        // At least one palace should have 忌
        let has_ji = a.palaces.iter().any(|p| p.has_mutagen(Mutagen::Ji));
        assert!(has_ji);
    }

    #[test]
    fn test_palace_is_empty() {
        let a = make_astrolabe();
        // Count non-empty palaces (those with StarType::Major stars)
        let non_empty = a.palaces.iter().filter(|p| !p.is_empty()).count();
        // Should have some non-empty palaces
        assert!(non_empty > 0);
        // Not all palaces can be non-empty (14 major stars across 12 palaces, some share)
        // but some palaces should be empty
        let empty_count = a.palaces.iter().filter(|p| p.is_empty()).count();
        // Total should be 12
        assert_eq!(non_empty + empty_count, 12);
    }

    #[test]
    fn test_palace_flies_to() {
        let a = make_astrolabe();
        let source = &a.palaces[0];
        // flies_to should work: the mutagen star from source's heavenly stem
        // should be found in some palace
        let mut found_target = false;
        for target in &a.palaces {
            if source.flies_to(target, Mutagen::Lu) {
                found_target = true;
                break;
            }
        }
        assert!(found_target, "Lu mutagen star should exist in some palace");
    }

    #[test]
    fn test_palace_self_mutaged() {
        let a = make_astrolabe();
        // self_mutaged checks if the palace's own heavenly stem's mutagen star is in itself
        // This may or may not be true for any given palace; just verify it doesn't panic
        for p in &a.palaces {
            let _ = p.self_mutaged(Mutagen::Lu);
            let _ = p.self_mutaged_one_of();
            let _ = p.not_self_mutaged();
        }
    }

    #[test]
    fn test_palace_mutaged_places() {
        let a = make_astrolabe();
        let places = a.palaces[0].mutaged_places(&a.palaces);
        assert_eq!(places.len(), 4);
        // Each mutagen star must exist in exactly one palace
        for place in &places {
            assert!(
                place.is_some(),
                "Each mutagen star should be found in some palace"
            );
        }
    }
}

File: python/x_iztro/palace.py
"""
宫位与大限。

宫位是星盘的基本单元：星耀落在宫位上，四化飞星在宫位之间发生，
大限与小限也按宫位排布。判断方法一律基于语言无关标识。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from x_iztro.enums import _MUTAGEN_INDEX, Mutagen, _as_mutagen_list, _or_all_mutagens
from x_iztro.star_object import Star, _star_identifiers

if TYPE_CHECKING:
    from x_iztro.astrolabe import Astrolabe
    from x_iztro.knowledge import KnowledgePack
    from x_iztro.surpalaces import SurroundedPalaces


@dataclass(frozen=True, slots=True)
class Decadal:
    """大限"""

    range: tuple[int, int]
    """大限起止年龄 (起始, 截止)"""

    heavenly_stem: str
    """大限天干（翻译文本）"""

    heavenly_stem_key: str
    """大限天干标识（`HeavenlyStem` 枚举值域）"""

    earthly_branch: str
    """大限地支（翻译文本）"""

    earthly_branch_key: str
    """大限地支标识（`EarthlyBranch` 枚举值域）"""

    @classmethod
    def _from_dict(cls, d: dict) -> Decadal:
        r = d["range"]
        return cls(
            range=(r[0], r[1]),
            heavenly_stem=d["heavenlyStem"],
            heavenly_stem_key=d["heavenlyStemKey"],
            earthly_branch=d["earthlyBranch"],
            earthly_branch_key=d["earthlyBranchKey"],
        )


# ============================================================
# Palace
# ============================================================

@dataclass(frozen=True, slots=True)
class Palace:
    """宫位"""

    index: int
    """宫位索引 (0-11)"""

    name: str
    """宫位名称（按排盘语言翻译）"""

    name_key: str
    """语言无关宫位标识（`PalaceName` 枚举值域）"""

    is_body_palace: bool
    """是否身宫"""

    is_original_palace: bool
    """是否来因宫"""

    heavenly_stem: str
    """宫位天干（翻译文本）"""

    heavenly_stem_key: str
    """宫位天干标识（`HeavenlyStem` 枚举值域）"""

    earthly_branch: str
    """宫位地支（翻译文本）"""

    earthly_branch_key: str
    """宫位地支标识（`EarthlyBranch` 枚举值域）"""

    major_stars: list[Star]
    """主星列表"""

    minor_stars: list[Star]
    """辅星列表"""

    adjective_stars: list[Star]
    """杂耀列表"""

    changsheng12: str
    """长生十二神（翻译文本）"""

    changsheng12_key: str
    """长生十二神标识（`Changsheng12` 枚举值域）"""

    boshi12: str
    """博士十二神（翻译文本）"""

    boshi12_key: str
    """博士十二神标识（`Boshi12` 枚举值域）"""

    jiangqian12: str
    """将前十二神（翻译文本）"""

    jiangqian12_key: str
    """将前十二神标识（`Jiangqian12` 枚举值域）"""

    suiqian12: str
    """岁前十二神（翻译文本）"""

    suiqian12_key: str
    """岁前十二神标识（`Suiqian12` 枚举值域）"""

    decadal: Decadal
    """大限信息"""

    ages: list[int]
    """小限经过年龄"""

    mutagen_star_keys: list[str]
    """本宫天干化出的四颗星标识，顺序为禄、权、科、忌。

    由排盘时生效的四化表算得，因此自定义四化表（`ChartConfig.mutagens`）
    会反映在这里，飞星判断随之改变。
    """

    _astrolabe: Astrolabe | None = field(default=None, compare=False, repr=False)
    """宫位所属星盘，由星盘构造时回填"""

    # ------ 星耀判断 ------

    def has(self, stars: list[str]) -> bool:
        """判断宫位是否包含指定的 **所有** 星耀（接受星耀枚举或当前语言的星名）"""
        identifiers = self._all_star_identifiers()
        return all(s in identifiers for s in stars)

    def not_have(self, stars: list[str]) -> bool:
        """判断宫位是否 **不** 包含指定的所有星耀"""
        identifiers = self._all_star_identifiers()
        return all(s not in identifiers for s in stars)

    def has_one_of(self, stars: list[str]) -> bool:
        """判断宫位是否包含指定星耀中的 **至少一颗**"""
        identifiers = self._all_star_identifiers()
        return any(s in identifiers for s in stars)

    def has_mutagen(self, mutagen: Mutagen) -> bool:
        """判断宫位是否有指定四化（只检查主星和辅星）"""
        return any(
            s.mutagen_key == mutagen for s in self.major_stars + self.minor_stars
        )

    def not_have_mutagen(self, mutagen: Mutagen) -> bool:
        """判断宫位是否没有指定四化"""
        return not self.has_mutagen(mutagen)

    def is_empty(self, exclude_stars: list[str] | None = None) -> bool:
        """
        判断宫位是否为空宫（无十四主星）。

        Args:
            exclude_stars: 视为「不空」的星耀；宫内出现其中任一颗即不作空宫论
        """
        if self.major_stars:
            return False
        if exclude_stars and self.has_one_of(exclude_stars):
            return False
        return True

    def astrolabe(self) -> Astrolabe | None:
        """宫位所属星盘；脱离星盘单独构造的宫位返回 None"""
        return self._astrolabe

    def to_text(self, *, knowledge: bool | KnowledgePack | None = None) -> str:
        """
        本宫的语义化文本：面向语言模型与人的单宫完整描述。

        从所属星盘的排盘上下文（含重排起点）无状态再发起计算，
        文本按排盘语言输出。

        Args:
            knowledge: 释义材料（True 取排盘语言的内嵌包，或给 KnowledgePack）；
                给出时追加本宫星耀的释义节

        Raises:
            ValueError: 宫位脱离星盘单独构造，无排盘上下文可转发
            IztroError: `knowledge=True` 而排盘语言没有内嵌包（目前只有 zh-CN）
        """
        if self._astrolabe is None:
            raise ValueError("to_text 需要所属星盘：请从 Astrolabe 的宫位查询获取宫位")
        return self._astrolabe._context_query(
            "palaceToText", palace_index=self.index, knowledge=knowledge
        )

    def opposite_palace(self) -> Palace | None:
        """
        本宫的对宫，即索引 +6 的那一宫；脱离星盘单独构造的宫位返回 None。

        对宫与本宫永远相对而看：命宫对迁移、财帛对福德，依此类推。
        """
        if self._astrolabe is None:
            return None
        return self._astrolabe.palaces[(self.index + 6) % 12]

    def surrounded_palaces(self) -> SurroundedPalaces | None:
        """本宫的三方四正；脱离星盘单独构造的宫位返回 None"""
        if self._astrolabe is None:
            return None
        return self._astrolabe.surrounded_palaces(self.index)

    # ------ 四化飞星 ------

    def mutagen_stars(self, mutagens: Mutagen | list[Mutagen]) -> list[str]:
        """
        本宫天干在指定四化位上对应的星耀标识。

        顺序与传入顺序一致；查不到对应星耀的四化位会被跳过。
        """
        out: list[str] = []
        for m in _as_mutagen_list(mutagens):
            idx = _MUTAGEN_INDEX.get(m)
            if idx is None or idx >= len(self.mutagen_star_keys):
                continue
            out.append(self.mutagen_star_keys[idx])
        return out

    def flies_to(
        self,
        target: Palace | int | str,
        mutagens: Mutagen | list[Mutagen],
    ) -> bool:
        """
        判断本宫天干的指定四化星是否 **全部** 落在目标宫位。

        Args:
            target: 目标宫位对象、宫位索引，或宫位标识/宫名
            mutagens: 四化类型，单个或列表

        四化列表为空时返回 False。
        """
        to = self._resolve_palace(target)
        if to is None:
            return False
        stars = self.mutagen_stars(mutagens)
        if not stars:
            return False
        return to.has(stars)

    def flies_one_of_to(
        self,
        target: Palace | int | str,
        mutagens: Mutagen | list[Mutagen],
    ) -> bool:
        """
        判断本宫天干的指定四化星是否有 **任意一颗** 落在目标宫位。

        四化列表为空时返回 True。
        """
        to = self._resolve_palace(target)
        if to is None:
            return False
        stars = self.mutagen_stars(mutagens)
        if not stars:
            return True
        return to.has_one_of(stars)

    def not_fly_to(
        self,
        target: Palace | int | str,
        mutagens: Mutagen | list[Mutagen],
    ) -> bool:
        """
        判断本宫天干的指定四化星是否 **一颗都不** 落在目标宫位。

        四化列表为空时返回 True。
        """
        to = self._resolve_palace(target)
        if to is None:
            return False
        stars = self.mutagen_stars(mutagens)
        if not stars:
            return True
        return to.not_have(stars)

    def self_mutaged(self, mutagens: Mutagen | list[Mutagen]) -> bool:
        """判断本宫天干的指定四化星是否 **全部** 落在本宫（自化）"""
        return self.has(self.mutagen_stars(mutagens))

    def self_mutaged_one_of(
        self, mutagens: Mutagen | list[Mutagen] | None = None
    ) -> bool:
        """
        判断本宫是否有指定四化中的任意一种自化。

        不传或传空表示检查全部四化。
        """
        return self.has_one_of(self.mutagen_stars(_or_all_mutagens(mutagens)))

    def not_self_mutaged(
        self, mutagens: Mutagen | list[Mutagen] | None = None
    ) -> bool:
        """
        判断本宫是否没有指定四化中的任何一种自化。

        不传或传空表示检查全部四化。
        """
        return self.not_have(self.mutagen_stars(_or_all_mutagens(mutagens)))

    def mutaged_places(
        self, all_palaces: list[Palace] | None = None
    ) -> list[Palace | None]:
        """
        查看本宫天干四化分别飞入哪些宫位。

        Args:
            all_palaces: 检索范围；不传则取本宫所属星盘的十二宫

        Returns:
            长度为 4 的列表 [禄飞入宫, 权飞入宫, 科飞入宫, 忌飞入宫]，
            未找到时为 None。
        """
        if all_palaces is None:
            if self._astrolabe is None:
                return []
            all_palaces = self._astrolabe.palaces

        result: list[Palace | None] = []
        for star_key in self.mutagen_stars(list(Mutagen)):
            result.append(next((p for p in all_palaces if p.has([star_key])), None))
        return result

    def _resolve_palace(self, target: Palace | int | str) -> Palace | None:
        """把宫位索引或宫名解析为宫位对象；宫位对象原样返回。"""
        if isinstance(target, Palace):
            return target
        if self._astrolabe is None:
            return None
        return self._astrolabe.palace(target)

    # ------ 内部方法 ------

    def _all_star_identifiers(self) -> set[str]:
        return _star_identifiers(
            self.major_stars + self.minor_stars + self.adjective_stars
        )

    @classmethod
    def _from_dict(cls, d: dict) -> Palace:
        return cls(
            index=d["index"],
            name=d["name"],
            name_key=d["nameKey"],
            is_body_palace=d["isBodyPalace"],
            is_original_palace=d["isOriginalPalace"],
            heavenly_stem=d["heavenlyStem"],
            heavenly_stem_key=d["heavenlyStemKey"],
            earthly_branch=d["earthlyBranch"],
            earthly_branch_key=d["earthlyBranchKey"],
            major_stars=[Star._from_dict(s) for s in d["majorStars"]],
            minor_stars=[Star._from_dict(s) for s in d["minorStars"]],
            adjective_stars=[Star._from_dict(s) for s in d["adjectiveStars"]],
            changsheng12=d["changsheng12"],
            changsheng12_key=d["changsheng12Key"],
            boshi12=d["boshi12"],
            boshi12_key=d["boshi12Key"],
            jiangqian12=d["jiangqian12"],
            jiangqian12_key=d["jiangqian12Key"],
            suiqian12=d["suiqian12"],
            suiqian12_key=d["suiqian12Key"],
            decadal=Decadal._from_dict(d["decadal"]),
            ages=list(d["ages"]),
            mutagen_star_keys=list(d["mutagenStarKeys"]),
        )
